crates/dbsp/src/profile/mod.rs

```diff
@@ -24,6 +28,294 @@ pub struct Profiler {
     circuit: RootCircuit,
 }
 
+/// Runtime profile of an individual DBSP worker thread.
+#[derive(Clone, Default)]
+pub struct WorkerProfile {
+    metadata: HashMap<GlobalNodeId, OperatorMeta>,
+}
+
+impl WorkerProfile {
+    fn new(metadata: HashMap<GlobalNodeId, OperatorMeta>) -> Self {
+        Self { metadata }
+    }
+
+    /// Returns the profile for a specific attribute.
+    ///
+    /// The returned hashmap contains id's of nodes that have the specified
+    /// attribute along with the value of the attribute.
+    pub fn attribute_profile(&self, attr: &str) -> HashMap<GlobalNodeId, MetaItem> {
+        let mut result = HashMap::new();
+
+        for (id, meta) in self.metadata.iter() {
+            if let Some(item) = meta.get(attr) {
+                result.insert(id.clone(), item);
+            }
+        }
+
+        result
+    }
+
+    /// Returns the profile for a specific attribute of type
+    /// [`MetaItem::Bytes`].
+    ///
+    /// Fails if the profile contains an attribute with the specified name and a
+    /// type that is different from [`MetaItem::Bytes`].  On error, returns
+    /// the value of the attribute that caused the failure.
+    pub fn attribute_profile_as_bytes(
+        &self,
+        attr: &str,
+    ) -> Result<HashMap<GlobalNodeId, HumanBytes>, MetaItem> {
+        let mut result = HashMap::new();
+
+        for (id, meta) in self.attribute_profile(attr).into_iter() {
+            if let MetaItem::Bytes(bytes) = meta {
+                result.insert(id, bytes);
+            } else {
+                return Err(meta);
+            }
+        }
+        Ok(result)
+    }
+
+    /// Returns the sum of values of an attribute of type [`MetaItem::Bytes`]
+    /// across all nodes.
+    ///
+    /// Fails if the profile contains an attribute with the specified name and a
+    /// type that is different from [`MetaItem::Bytes`].  On error, returns
+    /// the value of the attribute that caused the failure.
+    pub fn attribute_total_as_bytes(&self, attr: &str) -> Result<HumanBytes, MetaItem> {
+        Ok(HumanBytes::new(
+            self.attribute_profile_as_bytes(attr)?
+                .into_iter()
+                .fold(0u64, |acc, (_, item)| acc + item.bytes),
+        ))
+    }
+
+    /// Returns the profile for a specific attribute of type [`MetaItem::Int`].
+    ///
+    /// Fails if the profile contains an attribute with the specified name and a
+    /// type that is different from [`MetaItem::Int`].  On error, returns
```

Comment:
I don’t think of usize as int

Reply:
yeah, this is odd, but this isn't a new thing, so I didn't want to touch it in this PR.